Signal/src/view controllers/RegistrationViewController.m

```diff
@@ -148,48 +148,60 @@ - (void)dismissKeyboardFromAppropriateSubView {
+    // input as possible".
+    //
+    // * Always accept deletes.
+    // * Ignore invalid input.
+    // * Take partial input if possible.
+    
+    NSString *oldText = textField.text;
+    // Construct the new contents of the text field by:
+    // 1. Determining the "left" substring: the contents of the old text _before_ the deletion range.
+    //    Filtering will remove non-decimal digit characters like hyphen "-".
+    NSString *left = [oldText substringToIndex:range.location].digitsOnly;
+    // 2. Determining the "right" substring: the contents of the old text _after_ the deletion range.
+    NSString *right = [oldText substringFromIndex:range.location + range.length].digitsOnly;
+    // 3. Determining the "center" substring: the contents of the new insertion text.
+    NSString *center = insertionText.digitsOnly;
+    // 4. Construct the "raw" new text by concatenating left, center and right.
+    NSString *textAfterChange = [[left stringByAppendingString:center]
+                                 stringByAppendingString:right];
+    // 5. Construct the "formatted" new text by inserting a hyphen if necessary.
     // reformat the phone number, trying to keep the cursor beside the inserted or deleted digit
-    bool isJustDeletion = string.length == 0;
+    bool isJustDeletion = insertionText.length == 0;
+    NSUInteger cursorPositionAfterChange = left.length + center.length;
     NSString *textAfterReformat =
-        [PhoneNumber bestEffortFormatPartialUserSpecifiedTextToLookLikeAPhoneNumber:textAfterChange.digitsOnly
-                                                     withSpecifiedCountryCodeString:_countryCodeButton.titleLabel.text];
+    [PhoneNumber bestEffortFormatPartialUserSpecifiedTextToLookLikeAPhoneNumber:textAfterChange
+                                                 withSpecifiedCountryCodeString:_countryCodeButton.titleLabel.text];
     NSUInteger cursorPositionAfterReformat = [PhoneNumberUtil translateCursorPosition:cursorPositionAfterChange
                                                                                  from:textAfterChange
                                                                                    to:textAfterReformat
                                                                     stickingRightward:isJustDeletion];
     textField.text = textAfterReformat;
     UITextPosition *pos =
-        [textField positionFromPosition:textField.beginningOfDocument offset:(NSInteger)cursorPositionAfterReformat];
+    [textField positionFromPosition:textField.beginningOfDocument offset:(NSInteger)cursorPositionAfterReformat];
     [textField setSelectedTextRange:[textField textRangeFromPosition:pos toPosition:pos]];
-
+    
     return NO; // inform our caller that we took care of performing the change
 }
 
+- (BOOL)textFieldShouldReturn:(UITextField *)textField {
+    [self sendCodeAction:nil];
+    [textField resignFirstResponder];
```

Comment:
The return key now submits the phone number which will make it easier for developers to register in a simulator.